Recurring Payments: Replace submit button

aaronrobertshaw opened the pull request:
Fixes #15713

#### Changes proposed in this Pull Request:
* Replace the SubmitButton in the Recurring Payments block with new shared Button block.

#### Does this pull request change what data or activity we track or use?
No.

#### Prerequisites
* Sandboxed `public-api.wordpress.com` and `subscribe.wordpress.com`. See instructions for sandboxing for Recurring Payments in the Field Guide.
* Site connected to Stripe and a plan created.
* A post with a Recurring Payment block added prior to checking out this PR branch.

#### Testing instructions:

1. On the frontend, visit post containing previous version of Recurring Payments block.
2. Confirm the blocks display and function as they did when you set them up. 
    _Button should still be contained within div with `.wp-block-button` class._
3. Edit the post.
4. Confirm the Recurring Payment block still appears correctly and its settings are correct.
5. Update the Recurring Payment button using the new settings controls provided by button block.
6. Add new Recurring Payment block select a plan. Ensure the button text is updated with text including the price of the plan selected.
7. Save the post.
8. Re-visit the post on the frontend and confirm display and function of Recurring Payment block.

<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->
#### Before:

<img width="1031" alt="Screen Shot 2020-05-25 at 2 53 25 pm" src="https://user-images.githubusercontent.com/60436221/82779453-97aa6b80-9e97-11ea-8917-bbe1580edcfa.png">

#### After:

<img width="1032" alt="Screen Shot 2020-05-25 at 2 50 03 pm" src="https://user-images.githubusercontent.com/60436221/82779463-9c6f1f80-9e97-11ea-9b60-ddaea58440fc.png">


#### Proposed changelog entry for your changes:
* Update Recurring Payments block to use new shared Button block

aaronrobertshaw commented:
This block apparently allows for custom button HTML attributes from WPCOM VIP sites. I don't know how to test with VIP sites or even how their versions of this block might differ. 

If anyone can point me in the right direction of what needs doing, that would be great.

aaronrobertshaw commented on extensions/blocks/recurring-payments/edit.jsx:
```diff
@@ -504,6 +498,6 @@ class MembershipsButtonEdit extends Component {
 
 export default compose( [
-	withSelect( select => ( { postId: select( 'core/editor' ).getCurrentPostId() } ) ),
+	withSelect( selector => ( { postId: selector( 'core/editor' ).getCurrentPostId() } ) ),
```
Needed to import `select` from `@wordpress/data` to pass text attribute through to child inner blocks when the plan was selected. This meant the linter didn't like naming the arg here `select` as well, given it was already defined in the outer scope.

Open to suggestions on a better approach than this.

glendaviesnz replied:
is it possible to get innerButtons in the withSelect instead? ie.

```javascript

withSelect( ( select, { clientId } )  => ( { 
	postId: select( 'core/editor' ).getCurrentPostId(),
	innerButtons:  select( 'core/editor' ).getBlocksByClientId( clientId )	
} ) )

```

aaronrobertshaw commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -270,10 +307,10 @@ public function render_button( $attrs ) {
 			esc_attr( $button_styles ),
-			wp_kses( $data['button_label'], self::$tags_allowed_in_the_button ),
+			wp_kses( $button_label, self::$tags_allowed_in_the_button ),
 			isset( $attrs['submitButtonAttributes'] ) ? sanitize_text_field( $attrs['submitButtonAttributes'] ) : '' // Needed for arbitrary target=_blank on WPCOM VIP.
```
I'm not clear on where and how the `submitButtonAttributes` for WPCOM VIP sites are set. Can anyone shed any light on this and see if these changes might introduce other VIP related issues?

aaronrobertshaw replied:
Turns out these come from direct calls to `Jetpack_Memberships::get_instance()->render_button()`. As they won't supply inner block content, they'll be treated as deprecated buttons by `render_button`. 

I've removed the unnecessary handling of `submitButtonAttributes` while rendering the latest version of the block.

scruffian commented:
I've tested this and it works as advertised. I'll look at the code tomorrow 👍

aaronrobertshaw commented:
@glendaviesnz I've rebased this PR and resolved the conflict. 

It could use another quick test to have someone else confirm the custom button text remains when switching plans for the block. I've updated the test instructions and tested the block myself and it's working for me.

scruffian commented:
> It could use another quick test to have someone else confirm the custom button text remains when switching plans for the block. I've updated the test instructions and tested the block myself and it's working for me.

WFM 👍

fgiannar commented:
Hi @aaronrobertshaw ,

Tested this and all looks good, nice job!

My only issue is when aligning the button right, which causes the Payments' block properties to not appear anymore on click (but I believe this alignment issue is unrelated with the current PR, so if that's the case feel free to ignore it).

Also I see some build errors which will probably be resolved by rebasing against master.

aaronrobertshaw commented:
Hi @fgiannar ,

Thanks for testing this one. 

Not being easily able to select the parent block while the inner button block is aligned left or right is an issue. It is a side-effect of the switch to using inner blocks. 

That said, we do have the block navigation button in the main editor toolbar as well as the parent selector when hovering the button block’s toolbar.  There is also [currently work in progress](https://github.com/WordPress/gutenberg/issues/23766) ( [PR](https://github.com/WordPress/gutenberg/pull/23800) ) to improve the UX around the parent selector, making it more discoverable and obvious.

If you are happy with these options in the short term, I’d be inclined to proceed with this PR. What do you think?

jeherve commented:
r211585-wpcom
